### ISSUE TITLE
Alter how the Redirects' integrity check functions

### ISSUE THREAD

**Issue by phg49389** [betheluniversity/tinker#347]
Currently, on the first day of each month, Tinker iterates over all the redirects in its DB and makes GET requests to each "from path". After it finishes redirecting, it asserts that its final destination is the expected "to path". If they're different, then the "to path" gets updated and an email gets sent to Comm/Mktg. Additionally, it checks if a redirect path has expired. If it has, then it will automatically be removed from the DB table and a message gets added to the outbound email.

This is very good functionality, as it keeps the redirects file up-to-date and trimmed down to only what's active. However, because we have around 14,000 redirects, the cronjob is making 14k GET requests to the main site and that puts it under tremendous load. After some conversation, Caleb and I think we should change the functionality of the cronjob to this:

- The cronjob runs every day of the month, instead of only the first day
- When the Redirects' public endpoint is called, it gets the day of the month as an integer (1-31), as well as how many days are in the month (28-31). 
- It uses how many days are in the month to calculate how many redirects it should update, and which day of the month it is to determine what index it should start at. For example, on Feb 7, it would update the 7th section of 1/28th of all redirects in the DB. Ergo, it would update redirects 3001-3500. I recommend using the `ceil()` function on the division result.
- To account for redirect insertions and deletions each day, start 5 indexes before the calculated start and end 5 indexes after the calculated end. Obviously, this behavior should only apply to "middle days" (not the first day, not the last day). Perhaps something like `start_index = max(start_index - 5, 0)` and `end_index = min(end_index + 5, len(redirects)`?
- Once the start and end indexes have been calculated, then the current for loop should be easy to adapt to the new slice of the redirects array. Its functionality doesn't need to be changed in the regard.
- Finally, make sure that an email gets sent to Comm/Mktg _only_ if a redirect path gets updated or expires and gets deleted. This is because we don't to create cases every day for them that have nothing in them.

**Comment by c-schwarze** [comment]
@bak45247 this should be a good project for you to work on. Lets talk this through when you start working on this.

**Comment by c-schwarze** [comment]
We might even want to have this run less often. The last 3 times this ran (last 3 months), no notable updates were made.

**Comment by bak45247** [comment]
1 0 * * * path/to/redirect_check/script >> This is how we run everyday

use .len() to determine length of redirects list >> I'm not sure how this will work, will test and try again

use datetime to determine number of days in month (Only on the first of each month) (Either a separate cron or some extra logic inside of figuring out which day it is)

math.ceil(list.len() / days) to determine number of redirects done each day >> Again, not sure how .len() will work

datetime to figure which day it is currently

current * number daily - 5 = start point

start point + 10 + number daily = end point

run, append to result list, write result list + count of changing http > https

**Comment by bak45247** [comment]
work with a clock to run only a couple hundred at a time and delay
sleep (1 - 3 seconds to lighten load on server) (and not run forever)

email to send through tinker or python OR rework the bash script to only send the email if there are in fact changes to be made

tl;dr - 
Script will run everday, getting number of redirects to run based on the number of days in the month. Will run 1 and sleep for 1-3 seconds. We want to give the server time to breath, but not let the script run for more than an hour or so. 
With the email, we'll either send it through tinker with python OR change the logic in the script to not allow an email to be sent if the result list is empty.

**Comment by bak45247** [comment]
calendar.monthrange(datetime.year, datetime.month)[1] = THE method used for finding the number of days in a month = days_month

redirects.len() = the method for obtaining the total number of redirects = total_redirects

math.ceil(total_redirects / days_month) = number of redirects checked per day = redirects_per_day

if its not the 1st or the last day of the month:
starting_point = datetime.day * redirects_per_day - 5
ending_point = starting_point + redirects_per_day + 10
this way we start at 5 behind the actual start of the day and finish 5 beyond the endpoint of the day

1st day:
starting point = 0
ending point = starting_point + redirects_per_day + 5

last day:
starting_point = datetime.day * redirects_per_day - 5
end_point = total_redirects

this is how we calculate the start and stopping point at all days of the month

**Comment by bak45247** [comment]
All that is left is testing load on the server as well as changing the cron to run everyday

I don't have access to the server so I'm unable to change that. 

@c-schwarze if you could change the cron to have 0 1 * * * as the initial part so it will run daily taht would be great

**Comment by c-schwarze** [comment]
We still need 1 more change to be made, which is the email logic. Right now, it sends everytime it runs. That means we will email every day.

**Comment by bak45247** [comment]
Oh shoot yeah, can we look at the script on Wednesday when I come in?

**Comment by bak45247** [comment]
For the emails being sent, we're using flask_mail. 

Documentation: https://www.tutorialspoint.com/flask/flask_mail.htm

it doesn't actually work locally, so the local testing has just been assembling the emails and just printing them in the console. As far as true testing, I'll need to be on xp to be able to do that
